axum/src/docs/extract.md

```diff
@@ -251,6 +251,9 @@ let app = Router::new().route("/users", post(create_user));
 # };
 ```
 
+`Path` extractors have a specific behaviour and should only be used with `Result`
+wrapper, see [`path`](`Path`#optional-path-parameters) module doc for details.
```

Comment:
Hm I don't think I'd say that `PathRejection` is special. All rejections work like this. For example `Option<Json<T>>` doesn't tell you whether the `content-type: application/json` was missing or whether there was a deserialization error.

I think I'd just leave out this paragraph.

Reply:
Well, what I meant is that a missing path parameter will be rejected early and not reach the handler at all. Maybe I can rephrase this to emphasize that this is about rejection based on a route before any handler is called

Reply:
But that is also not special for `Path`. Any rejection will cause the handler to not be called.

Reply:
I seem to misunderstand how this works, but I can't get a clear picture. Let's take the optional extractor example, there we have `Option<Json<Value>>` and missing json does not mean that we will get a rejection without calling the handler.

In the case of Path extractor, on the other hand, the parameters should match the route path first and only then they are passed to handler, that might also reject the result based on its own parameters.

Reply:
> In the case of Path extractor, on the other hand, the parameters should match the route path first and only then they are passed to handler, that might also reject the result based on its own parameters.

Thats not correct. If you do `.route("/", get(|params: Option<Path<String>>| async {}))`, i.e a route without any path params using `Option<Path<_>>` then the handler is still always called. `params` will always be `None`.

A mismatch between the route and `Path` triggers a rejection which `Option` discards and yields a `None` (https://github.com/tokio-rs/axum/blob/main/axum-core/src/extract/mod.rs#L153).

Reply:
So `Option<Json<T>>` being None is not a rejection, or is that something else I am missing?

Reply:
What do you mean by " is not a rejection"? `Option<Json<_>>` is an extractor whose rejection is `Infallible` (ie never rejects)

Reply:
I feel like you're misunderstanding what I'm trying to say.

> if I do `.route("/", get(|params: Option<Path<String>>| async {}))` it will handle the missing path parameter as expected, because there is no parameter in the path, nowhere to get it from

In this case the handler will _always_ be called and `params` will be `None`. No rejections will happen.

> if I do `.route("/:parameter/suffix", get(|params: Option<Path<String>>| async {}))` then there is no way to get this handler called

Here the handler will also _always_ be called and `params` will be a `Some`. Again no rejections. If the route matches then the path parameter can be extracted and `String` cannot fail deserialization.

> even if I do `.route("/:parameter", get(|params: Option<Path<String>>| async {}))` it will end up rejected both without trailing slash, and with it (because with trailing slash there will be a redirection ending with rejection, too)

Same here as the previous example. It'll always work.

Reply:
I was talking about those routes being called without parameter, like `//suffix`

well, anyway, maybe this issue will at least pop up in search for someone. This week I've already explained to a colleague why they shouldn't use `Option<Path>` directly in axum handler and rather create a shim like you showed, so this is not something that's not going to be used just because it is not clearly mentioned in the docs.

I'd argue that it would be better to change the relevant part as per your comments to show the way it should be used rather than have everyone come up with their own ideas of what to do in such a case